addons/mass_mailing/models/mailing.py

```diff
@@ -266,7 +266,10 @@ def _compute_next_departure(self):
         for mass_mailing in self:
             if mass_mailing.schedule_date:
                 schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                if fields.Datetime.now() < schedule_date < cron_time:
+                    mass_mailing.next_departure = schedule_date
+                else:
+                    mass_mailing.next_departure = max(schedule_date, cron_time)
```

Comment:
Neat, didn't know about comparison chaining.

Not saying we can't, but isn't this just lying to users to make the interface feedback less confusing?

From the PR message I get the impression this is meant to be more accurate, but if the cron runs after the scheduled date, then that's when the mailing would be sent. So it's less accurate, right?
Unless I'm missing something

Reply:
Hey thanks for taking a look!

Since Odoo 15 ( might even been added in 14.2) we have the cron triggers which allow for a cron job to be triggered before its saved nextcall value.

From my naive understanding, when the shared cron workers on a SAAS server show up to that specific DB, they check if there is a trigger with a datetime before the `nextcall`, and runs the job early. There seems to be a time buffer where the job might start slightly later (the workers might be busy handling other jobs), but it should be close to the datetime of the trigger.

The default value for the “Mail Marketing: Process queue” is once every 1 days. so lets say:
- User clicks send now or schedules the mailing list for FEB10 9H10
- Trigger is created for that datetime
- Nextcall for “Mail Marketing: Process queue” happens to be FEB11 10H00

The user will see a banner with "This mailing is planned for FEB11 10H00". This might be confusing and lead the user to think that there was a mistake. Worse in my support ticket case he taught it was a good idea to go into the technical menu and run  “Mail Marketing: Process queue” manually, which lead to a bug where the mailing campaign was sending mails twice to the same recipients (out of scope of this precise PR).

So while the cron trigger does not guarantee that the job will run at that precise moment, is is closer to the truth than informing the client that the mails will go out 1 day later vs an hour later (based on the trigger).

Reply:
Oh wow, I see. Thanks for the explanation, I naively assumed nextCall accounted for these.
Indeed this looks like it was not updated when the time-scheduled cron job was added to mailings.

In that case why not do something like max(fields.Datetime.now(), schedule_date)? If I schedule in the past, it will be processed 'instantly' (on the next pass of a worker on this DB), and otherwise it will always be triggered around the scheduled date with the new cron trigger, right?

Reply:
When you breakpoint the `def _compute_next_departure(self):` and then use the front-end to schedule a "draft" mailing list, it runs a few times over that method.
- First run, since `mass_mailing.schedule_date` is not set (FALSE), it defaults to `mass_mailing.next_departure = cron_time`
- Further down stream, `action_schedule_date` takes the user input for a date, sets `mass_mailing.schedule_date` and creates the cron.trigger at the same date.time (using `self.mass_mailing_id.action_put_in_queue` in [mailing_mailing_schedule_date.py](https://github.com/odoo/odoo/blob/16.0/addons/mass_mailing/wizard/mailing_mailing_schedule_date.py))
- It comes back to `_compute_next_departure` and triggers the `if mass_mailing.schedule_date` condition. Only here my changes apply for the described edge case.

So your simplification would be to assume that the cron trigger (implicitly created) time is more relevant than cron time and just replace `max(schedule_date, cron_time)` with `max(fields.Datetime.now(), schedule_date)` ?

I can get behind that, since the behavior since Odoo 15 should be the one described above. I just hope that I am not missing some unforeseen consequences of that change.

Reply:
Tested `max(fields.Datetime.now(), schedule_date) ` : behaves as expected in my local test and seems a clean solution.
Pushed a new commit. Do you have any other input or ideas?